src/liboslcomp/typecheck.cpp

```diff
@@ -1180,58 +1156,422 @@ ASTfunction_call::typecheck_struct_constructor ()
+            m_compiler->errhandler().message("  Candidates are:\n");
     }
 
-    // Now look for a coercible match of args, exact march on return type
-    if (! match) {
-        m_typespec = typecheck_all_poly (expected, true, false);
-        if (m_typespec != TypeSpec())
-            match = true;
+    void reportAmbiguity(FunctionSymbol* sym) const {
+        int advance;
+        const char *formals = sym->argcodes().c_str();
+        TypeSpec returntype = m_compiler->type_from_code (formals, &advance);
+        formals += advance;
+
+        auto& errh =  m_compiler->errhandler();
+        errh.message("    ");
+
+        if (ASTNode* decl = sym->node())
+            errh.message("%s:%d\t", decl->sourcefile(), decl->sourceline());
+
+        errh.message("%s %s (%s)\n", m_compiler->type_c_str(returntype),
+                     sym->name(),
+                     m_compiler->typelist_from_code(formals).c_str());
     }
 
-    // Now look for a coercible match of args, equivalent march on return type
-    if (! match) {
-        m_typespec = typecheck_all_poly (expected, true, true);
-        if (m_typespec != TypeSpec())
-            match = true;
+    std::pair<FunctionSymbol*, TypeSpec> best(ASTNode* caller, bool strict = 0) {
+        switch (m_candidates.size()) {
+            case 0:  return { nullptr, TypeSpec() };
+            case 1:  return { m_candidates[0].sym, m_candidates[0].rtype };
+            default: break;
+        }
+
+        int ambiguity = -1;
+        std::pair<const Candidate*, int> c = { nullptr, -1 };
+        for (auto& candidate : m_candidates) {
+            // re-score based on matching return value
+            if (candidate.rscore > c.second) {
+                ambiguity = -1; // higher score, no longer ambiguous
+                c = std::make_pair(&candidate, candidate.rscore);
+            } else if (candidate.rscore == c.second)
+                ambiguity = candidate.rscore;
+        }
+
+        if ((ambiguity != -1) || strict) {
+            ASSERT (caller);
+            reportError(caller, m_candidates[0].name(), false,
+                        !m_candidates.empty(), "Ambiguous call to");
+            for (auto& candidate : m_candidates) {
+                if (candidate.rscore >= ambiguity)
+                    reportAmbiguity(candidate.sym);
+            }
+        }
+
+        ASSERT (c.first);
+        return {c.first->sym, c.first->rtype};
+    }
+
+    bool empty() const { return m_candidates.empty(); }
+};
+
+
+/// <doc LegacyOverload>
+///
+/// Check how a polymorphic function variant was chosen in prior versions.
+/// Check is performed based on the env variable OSL_LEGACY_FUNCTION_RESOLUTION
+///
+/// OSL_LEGACY_FUNCTION_RESOLUTION      // check resolution matches old behavior
+/// OSL_LEGACY_FUNCTION_RESOLUTION=0    // no checking
+/// OSL_LEGACY_FUNCTION_RESOLUTION=err  // check and error on mismatch
+/// OSL_LEGACY_FUNCTION_RESOLUTION=use  // check and use prior on mismatch
+/// </script>
```

Comment:
You don't have to worry about the docdeep format for internals, that's just for certain external APIs that we want to make doc web pages about.

You don't want the `</script>` or `~~~~`. Those do something different than you think; they were specifically for fully quoting large code sections that followed.